File: 1705/main.go
package main

import (
	"container/heap"
	"fmt"
)

func main() {
	fmt.Println(eatenApples([]int{1,2,3,5,2}, []int{3,2,1,4,2}))
}

func eatenApples(apples []int, days []int) int {
	h := hp{}
	ans := 0
	i := 0
	for ; i < len(apples); i++ {
		for len(h) > 0 && h[0].end <= i {
			heap.Pop(&h)
		}
		if apples[i] > 0 {
			heap.Push(&h, pair{i+days[i], apples[i]})
		}
		if len(h) > 0 {
			h[0].left--
			if h[0].left == 0 {
				heap.Pop(&h)
			}
			ans++
		}
	}
	for len(h) > 0 {
		for len(h) > 0 && h[0].end <= i {
			heap.Pop(&h)
		}
		if len(h) == 0 {
			break
		}
		p := heap.Pop(&h).(pair)
		num := min(p.end-i, p.left)
		ans += num
		i += num
	}
	return ans
}

type pair struct {end, left int}
type hp []pair

func (h hp) Len() int {return len(h)}
func (h hp) Less(i, j int) bool {return h[i].end < h[j].end}
func (h hp) Swap(i, j int) {h[i], h[j] = h[j], h[i]}
func (h *hp) Push(v interface{}) {*h = append(*h, v.(pair))}
func (h *hp) Pop() interface{} { a := *h; v := a[len(a)-1]; *h = a[:len(a)-1]; return v}

File: 1705/main.rs
use std::collections::BinaryHeap;
use std::cmp::Reverse;

struct Solution{}

impl Solution {
    pub fn eaten_apples(apples: Vec<i32>, days: Vec<i32>) -> i32 {
        let mut ans = 0;
        let mut pq = BinaryHeap::new();
        let n = apples.len();
        let mut i = 0;
        while i < n {
            while let Some(Reverse((rotten_day, _count))) = pq.peek() {
                if *rotten_day <= i as i32 {
                    pq.pop();
                } else {
                    break;
                }
            }
            let rotten_day = i as i32 + days[i];
            let _count = apples[i];
            if _count > 0 {
                pq.push(Reverse((rotten_day, _count)));
            }
            if let Some(Reverse((rotten_day, mut _count))) = pq.pop() {
                _count -= 1;
                if _count > 0 {
                    pq.push(Reverse((rotten_day, _count)));
                }
                ans += 1;
            }
            i += 1;
        }
        while let Some(Reverse((rotten_day, c))) = pq.pop() {
            if rotten_day <= i as i32 {
                continue
            }
            let num = std::cmp::min(rotten_day - i as i32, c);
            ans += num;
            i += num as usize;
        }
        ans
    }
}

fn main() {
    println!("{}", Solution::eaten_apples(vec![1,2,3,5,2], vec![3,2,1,4,2]));
}